File: trees/543.py
# Tc = O(n)
# Sc = O(1) + Recursion Stack(max tree depth -> O(n)) => O(n)
class Solution:
    def solve(self, root, result):
        if root is None:
            return 0
        left = self.solve(root.left)
        right = self.solve(root.right)
        result[0] = max(result[0], left + right)
        return max(left, right) + 1
    def diameterOfBinaryTree(self, root):
        if root is None:
            return 0
        result = [float('-inf')]
        self.solve(root, result)
        return result

File: trees/543.js
class Solution{
    solve(root, result){
        if(root === null){
            return 0
        }
        const left = this.solve(root.left, result)
        const right = this.solve(root.right, result)
        result[0] = Math.max(result[0], left + right)
        return max(left, right) + 1
    }
    diameterOfBinaryTree(root){
        if(root === null){
            return 0
        }
        const result = [Number.NEGATIVE_INFINITY]
        this.solve(root,result)
        return result[0]
    }
}